### PR TITLE
One projection pipeline and the Direct execution adapter

### PR DESCRIPTION
**Pull request by habdelra**
Every execution tier is supposed to answer with the same records for the same card. The way that stops being true is not a disagreement anyone can see — it is a second builder. Competing builders each agree with themselves, so nothing goes red until two of them are compared, and by then every consumer has grown a preference for one of the answers.

So this lands exactly one pipeline, in two halves.

**The capture** (`boxel-projection.ts`) is everything that touches a live object: the Card API, the Loader, an instance's own getters. It reads a type or an instance once and answers with values — a plain graph of strings, numbers, booleans, nulls, arrays and plain objects. One capture is what makes a record internally coherent: an instance's field list and its model come from a single read, so a card whose fields and values were read at different moments never describes a state the instance was never in. Each operation captures the one record it answers with, so none of them pays for the other two.

**The assembler** (`boxel-render-record.ts`) is pure. Its whole input is what the capture produced, so given one capture the records are a function of that capture alone. It is where two guarantees get established once instead of at every call site: the protocol envelope, and inertness — each record is rebuilt through the protocol's own normalizer, so an accessor, a class instance, a symbol-keyed member, or a value containing itself is refused where it was produced, with the offending member's origin still in the stack, rather than at the far side of a message port where the diagnostic names only a path.

The rules the pipeline holds are the ones the spec names. Linked cards cross as `{$boxel:{id,type}}` identity and nothing else — a card linking a card linking a card hands its recipient no graph to walk. Contains composites expand in place, because they are embedded data and not a separate resource. Only a `present` link slot yields a reference: the other four states have a reference string but no loaded value and therefore no *actual* class, and every trust decision follows an instance's actual class rather than a field's declared type. Configuration resolves against the owning instance, and a type reports none rather than an unresolved one, because resolution has no `this` to run with. A themed card's theme members are derived Host-side, since resolving the linked Theme is precisely the graph walk a projection forbids — by calling Base's own theme helpers, which this PR lifts out of `getBoxComponent` and exports rather than reimplementing, so there is one spelling of that derivation in the tree. A trusted Base value's plain prototype getters are evaluated and carried, because a plain getter is not a field and a tier holding only the field-derived view of a currency amount renders a different card.

**`DirectBoxelRuntime`** implements the runtime contract over today's trusted Loader and Card API. It is the reference implementation, not a bypass: everything it answers with goes through that one pipeline, so a semantic the operations cannot express is an incomplete interface — and here, in the tier that holds the live objects and could cheat, is where that has to be found out, before a tier that cannot cheat is built against the gap. It loads through the Host's own Loader, so a module reaches it already through `transpileJS()` with its scoped CSS delivered by that pipeline's side-effect import; there is no second stylesheet compiler here. It retains the canonical Store instance rather than re-deserializing one, which would produce an object with the same data and none of the identity, lazy relationships, or mutation context. Rendering sits beside the interface rather than in it, because a mountable component is process-local and not cloneable.

`purpose` reaches no decision in Direct, and that is the honest answer for this tier rather than an omission — main fails loudly on a definition it cannot identify whatever the caller wanted it for, and there is no purpose Direct could be lenient for. A test pins that: the same materialization rejects under `host-display` and under `indexing`.

**A development-only diagnostic** rounds it out. A member the pipeline failed to project is not an error anywhere — it reads as `undefined`, the binding renders empty, and the card is subtly wrong with nothing in any log. `watchProjectionPaths` hands back a view that reports the complete path, the type, the format, and the execution mode when a consumer reads a path the record does not carry. It never synthesizes a value, and in a production build it is the identity function, so no proxy exists and no read is intercepted. It is offered rather than applied to what `projectInstance` returns: a record that crossed a boundary as a wrapper would be neither cloneable nor comparable.

**Two changes outside the pipeline.** `FieldDescription` and `ResolvedField` gain `isQueryBacked`, and the spec is amended to match: render-time writability is `(not computeVia) ∧ (not queryDefinition) ∧ permissions.canWrite`, the permissions term is context the Host pushes per surface, and a record carrying only `isComputed` cannot state the rule at all — a query-backed relationship is never editable and is not computed. `resolveFieldConfiguration` becomes a public Card API export, which is what lets the capture resolve configuration through the module the Loader served rather than a copy of its merge semantics.

The tier registry now records the adapter's file against the `direct` mode, and both the unit and the integration suite hand Direct's records to `checkRecordParity` with `direct` registered — which with one tier still checks something real, since every record it is given is read as data.

Nothing renders through any of this. No call site changed.

**Testing.** All green locally against a dev stack: 21 integration tests over a real realm module, a real Store-resident instance, and real links (62 assertions) — covering the theme derivation, a trusted Base getter reached through an authored subclass, a query-backed field, and that a projection triggers no load of its own; 10 unit tests on the assembler; 6 on the diagnostic; and the 130 existing rendering-protocol unit tests, whose fixtures pick up the new field member. `lint:tier-registry`, `lint:protocol-closure`, `lint:rp-bijection`, host `lint:types` and `eslint` all pass. The bijection's uncovered-statement count drops from 104 to 88.

**Three interface gaps this surfaced**, each recorded rather than papered over, because finding them here is what building Direct first is for. Nested-composite field configuration is not addressable through the operations at all: `getFields` answers a flat list keyed by a root-level name, so a tier rendering `venue.name` cannot obtain its `@configuration`. A tier that renders from records has nothing to trigger a lazy load with, since it never reads a getter on the canonical instance — so the rule that the renderer is the trigger holds for Direct and not beyond it. And query-backed whole-field states have no spelling in the record: a search that has not started projects the same empty membership as one that matched nothing. All three want protocol changes rather than adapter workarounds, and are tracked for the tickets that own those records.

### CODE REVIEW

File: packages/host/app/lib/boxel-projection.ts [hunk 1]
```diff
@@ -0,0 +1,837 @@
...
+  api: CardAPIModule,
+): (BoxelValueReference | null)[] {
+  void (instance as unknown as Record<string, unknown>)[fieldName];
+  let { membership } = api.getRelationshipMembershipState(
+    instance as CardDef,
+    fieldName,
+  );
+  return (membership ?? []).map((slot) =>
+    slot.kind === 'present' ? boxelReference(slot.value) : null,
+  );
```
[the remark this answers is not in the document]
Comment (habdelra) [reply packages/host/app/lib/boxel-projection.ts:557]: [Claude Code 🤖] Right, and it was worse than a null id — the two unsaved slots of a plural field were indistinguishable from each other. `slot.reference` is the target's id once saved and its local id before then, and both resolve through the Store's identity map, so `boxelReference` now takes it as a fallback identity. Fixed in 1474e1efd0.

The root instance's own `id` is left null when unsaved, which I think is the honest answer there — the instance genuinely has no id, and the protocol has no spelling for a local one. If a consumer needs to address an unsaved root we should add that member deliberately rather than overload `id`.

File: packages/host/app/lib/boxel-projection.ts [hunk 2]
```diff
@@ -0,0 +1,837 @@
...
+function captureTrustedGetters(
+  value: BaseDef,
+  expanded: Record<string, JsonValue>,
+): void {
+  let prototype = Object.getPrototypeOf(value) as object | null;
+  while (prototype && prototype !== Object.prototype) {
+    let owner = (prototype as { constructor?: BaseDefConstructor }).constructor;
+    let ref = owner ? identifyCard(owner) : undefined;
+    if (!ref || !isTrustedModule(moduleFrom(ref))) {
+      return;
```
[the remark this answers is not in the document]
Comment (habdelra) [reply packages/host/app/lib/boxel-projection.ts]: [Claude Code 🤖] Confirmed, and the comment two lines above claimed the behavior you describe as correct while the code did the opposite — an authored subclass contributed nothing at all. The walk now steps over an untrusted prototype instead of stopping.

One refinement on the suggestion: names the authored class *defines* are recorded and skipped for the trusted ancestors too. Otherwise an author who overrides `symbol` gets Base's value silently substituted underneath their own, which is the same class of divergence pointing the other way. Result is the author's members absent and the trusted members they did not touch present. Fixed in 1474e1efd0, with a test over a real `CurrencyField`.

File: packages/host/app/lib/direct-boxel-runtime.ts [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+  async getFields(
+    boxel: BoxelTypeHandle | BoxelInstanceHandle,
+  ): Promise<ResolvedField[]> {
+    let api = await this.getCardAPI();
+    if (this.types.has(boxel)) {
+      return buildResolvedFields(
+        captureUnresolvedFields(this.types.get(boxel), api),
+      );
+    }
+    return (await this.renderRecordFor(this.instances.get(boxel))).fields;
```
[the remark this answers is not in the document]
Comment (habdelra) [reply packages/host/app/lib/direct-boxel-runtime.ts]: [Claude Code 🤖] Agreed — `getFields`/`getField` now build from `captureBoxelFields` directly. Fixed in 1474e1efd0.

Worth adding that it had a second effect you didn't name: `renderRecordFor` also incremented `revision`, so a metadata read counted as a projection. Since no record path is declared tier-specific, `revision` is compared by the parity harness — a tier whose `getField` quietly counted and one whose didn't would diverge on a member neither controls. Only `projectInstance` increments it now, and the comment says why.

File: packages/host/app/lib/boxel-projection.ts [hunk 4]
```diff
@@ -0,0 +1,1004 @@
...
+    return Number.isFinite(value) ? value : null;
+  }
+  if (value instanceof URL) {
+    return value.href;
+  }
+  if (value instanceof Date) {
+    return value.toISOString();
+  }
+  if (typeof value !== 'object') {
+    return notData;
```
Comment (backspace) [inline packages/host/app/lib/boxel-projection.ts:957]: [Claude Code 🤖] `bigint` falls through this ladder to `notData`, and `captureValue` maps `notData` to `null` — so a `BigIntegerField` (`static [primitive]: bigint`, `packages/base/big-integer.gts`) projects as `null` on every tier, with nothing anywhere saying so. `watchProjectionPaths` cannot catch it either: it reports paths the record *lacks*, and `model.<field>` is present, holding `null`.

Two asks, separable.

`bigint` wants a spelling here, the way `URL` and `Date` two lines up have one — `BigIntegerSerializer`'s decimal string is what every other view of that value already uses. (`ResponseField`'s `Response` is the other non-JSON `static [primitive]` in Base; that one presumably *should* drop, which is an argument for making the drop visible rather than for widening the ladder.)

And `notData` reaching `captureValue` should be observable. The sentinel's own docblock says it exists so a caller "can tell 'this could not cross' from 'this is absent' and from 'this is null'", and `captureValue` is the one call site where that distinction has consequences — it discards it. `captureTrustedGetters` already does the alternative: omit the member. Omitting it here would route the read into the diagnostic's missing-path report instead of a silent `null`.

Regression, non-blocking — but the second half is worth settling before a consumer exists, because a tier holding the record has no way back to the difference.

**Review by backspace — COMMENTED** [review]
[Claude Code 🤖] This review went after the capture half's fidelity to what main actually reads, the twin implementations the pipeline introduces, and the type boundaries the `as` casts cross. It did not run the adapter against a realm stack or re-derive the spec statements from the doc.

**No blocking issues.** Five things to settle, all non-blocking. The one worth settling before a consumer exists is the silent `null` for a value that could not cross — a tier holding the record cannot tell it from an empty field, and the diagnostic built for exactly that class structurally cannot see it.

From the earlier round: the link-identity fallback and the trusted-getter walk both hold under the paths I traced. The `getFields` fix removes the cost it named, and I've continued in that thread — the same cost still stands on the other side of it, and the module header now asserts an invariant the fix removed.

1. `bigint` has no spelling in `dataValue`, so a `BigIntegerField` projects as `null`, and `notData` is collapsed to `null` at the one call site the sentinel exists for — thread on `dataValue` in `boxel-projection.ts`.
2. A query-backed link's never-queried / in-flight state projects as `[]`, identical to a resolved-empty search; a whole-field search failure projects as `[null]` — thread on `linkSlots`.
3. `captureTheme` re-implements five module-private helpers from `field-component.gts`, and the two copies already differ — thread on `captureTheme`.
4. `serializeCard`'s `as never` crosses `BaseDef` → `CardDef`, which `retainInstance` and `boxelKindOf` both make reachable — thread on `serializeCard` in `direct-boxel-runtime.ts`.
5. `projectInstance` builds and discards a full description and field list — in the existing `getFields` thread.

File: packages/host/app/lib/boxel-projection.ts [hunk 5]
```diff
@@ -0,0 +1,1004 @@
...
+  instance: BaseDef,
+  fieldName: string,
+  api: CardAPIModule,
+): (BoxelValueReference | null)[] {
+  let { membership } = api.getRelationshipMembershipState(
+    instance as CardDef,
+    fieldName,
+  );
+  return (membership ?? []).map((slot) =>
+    slot.kind === 'present'
```
Comment (backspace) [inline packages/host/app/lib/boxel-projection.ts:548]: [Claude Code 🤖] `getRelationshipMembershipState` answers `membership: undefined` for a query-backed field that is in flight or was never queried — the probe supplies `queryMembership` only once the search resource exists and has settled. `?? []` turns all of that into `[]`, which is also what a search that resolved to nothing produces. A tier iterating `@model.<queryField>` renders "no results" for a query that has not run.

The same read has a second case: a search that failed as a unit surfaces one whole-field sentinel, so `membership` is a one-element array whose slot is not `present` — which maps to `[null]`. A template iterating that renders one empty row where main renders the broken-link presentation.

The docblock enumerates RP-7.1's five *slot* states carefully and reaches neither of these, because both are whole-field states. Worth deciding what a query-backed field projects across never-queried / in-flight / resolved-empty / failed and saying so; `isLoading` is already destructured off this call and dropped.

The integration fixture has a query-backed `revivals`, and it is only ever asserted through `getFields` — nothing asserts `model.revivals`.

Regression, non-blocking.

File: packages/host/app/lib/boxel-projection.ts [hunk 6]
```diff
@@ -0,0 +1,1004 @@
...
+ * carries only as a reference — resolving that reference is exactly the graph
+ * walk a projection forbids. So the Host resolves it once, here, and the
+ * derived strings cross as data. Without them a themed card renders unthemed,
+ * which is not a degraded theme but a different design.
+ *
+ * The branch reproduces `field-component.gts`'s own `isThemeCard` / `themeCss`
+ * / `hasTheme` / `themeId` / `getCssImports`, which are module-private there. A
+ * card declaring its own `cssVariables` through a `CSSField` is a Theme card
+ * and scopes to its own identity; every other card scopes to the Theme its
+ * `cardTheme` mirror links.
```
Comment (backspace) [inline packages/host/app/lib/boxel-projection.ts]: [Claude Code 🤖] Five closures re-implemented, with nothing to go red when the two copies drift — which is this file's own header argument one level down: competing builders each agree with themselves.

They already differ. Main's `getCssImports` hands `card.cssImports` back untouched; `cssImportsOf` filters non-strings and maps an empty array to `null`. Harmless today, and exactly the kind of difference that has no test on either side.

This PR already established the fix in the neighbouring case: `resolveFieldConfiguration` became a public Card API export so the capture could resolve configuration through the module the Loader served rather than a copy of its merge semantics. The theme helpers want the same treatment — they are closures inside `getBoxComponent`, so lifting them to module scope in `field-component.gts` (or a small shared module) and exporting them is the move, and `captureTheme` then derives from main's own answers.

Follow-up if the export is more than this PR should carry — but then say so in the docblock, because "reproduces" currently reads as a design choice rather than a debt.

File: packages/host/app/lib/direct-boxel-runtime.ts [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+   * instance holds — serializing it would freeze a snapshot into
+   * `relationships`, and a consumer materializing that document would get a
+   * declared link where main has a query: editable where RP-7.6 says never,
+   * and stale the moment the search moves.
+   */
+  async serializeCard(
+    instance: BoxelInstanceHandle,
+  ): Promise<LooseSingleCardDocument> {
+    let api = await this.getCardAPI();
+    return api.serializeCard(this.instances.get(instance) as never, {
```
Comment (backspace) [inline packages/host/app/lib/direct-boxel-runtime.ts]: [Claude Code 🤖] `as never` is crossing `BaseDef` → `CardDef`, and the gap is reachable: `retainInstance` takes a `BaseDef`, the instance registry is `RuntimeHandleRegistry<BaseDef>`, and `boxelKindOf` explicitly answers `'field'`. `api.serializeCard` reads `model[localId]`, which `CardDef` declares and `FieldDef` does not — so a `FieldDef` handle reaching here builds `{type: 'card', lid: undefined}` and dies inside `isSingleCardDocument` with a shape dump, rather than saying the handle names a field.

Either narrow the instance registry to `CardDef` (and `retainInstance` with it), or check here and throw the named error. The cast is currently the only thing standing between the two types.

Regression, non-blocking.

File: packages/host/app/lib/direct-boxel-runtime.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (backspace) [reply packages/host/app/lib/direct-boxel-runtime.ts]: [Claude Code 🤖] The `getFields` / `getField` half lands — both build from `captureBoxelFields` now and no longer evaluate the model. Two things the change left standing, on the other side of it.

`projectInstance` still goes through `renderRecordFor`, which captures the type description and the resolved field list, normalizes all three records, and returns `.projection`. Nothing reads the other two: `buildBoxelRenderRecord` has one production caller, `renderRecordFor` has one caller, and it reads one member. So every projection runs the ancestry and format walks, resolves configuration for every field (evaluating authored `configuration` functions), and pays two extra `normalizeJsonData` passes for members no caller can reach — the argument the new comment above makes, in the other direction.

And `boxel-projection.ts`'s header still says one capture "keeps `getFields` and `projectInstance` from being able to disagree about one instance — both are built from a single read of it." That stopped being true with this fix: `getFields` now does its own read at its own moment. Each record stays internally coherent; the two *operations* are no longer built from one read of the instance. So either the header narrows to the claim that survives, or the split wants a different shape.

Concretely: `projectInstance` can capture the instance half alone and call `buildInstanceProjection` directly, leaving `buildBoxelRenderRecord` for a caller that wants all three.

File: packages/host/app/lib/boxel-projection.ts (hunk 4 above)
Comment (habdelra) [reply packages/host/app/lib/boxel-projection.ts:957]: [Claude Code 🤖] Both, and the second was the more useful half. Fixed in d69d0b43c4.

`bigint` has a spelling now — `String(value)`, which is what `BigIntegerSerializer.serialize` writes, so a projected value and a serialized one read the same.

And `notData` no longer collapses to `null` at a member: a field whose value cannot cross is **absent** from the model, so reading it routes through the missing-path diagnostic instead of looking like a field that genuinely is null. One deliberate exception — a `containsMany` element keeps `null`, because a position is an element's identity to anything iterating it and omitting one would renumber its siblings. That leaves `ResponseField` as the case that drops, visibly.

File: packages/host/app/lib/boxel-projection.ts (hunk 5 above)
Comment (habdelra) [reply packages/host/app/lib/boxel-projection.ts:548]: [Claude Code 🤖] Confirmed both, and the docblock did claim coverage it did not have. Addressed in d69d0b43c4 as documentation plus a pinning test rather than a shape change.

`membership: undefined` and the whole-field failure sentinel are exactly that — *whole-field* states — and RP-7.1's union describes slots, so neither has a place in it and `isLoading` has nowhere in the record to go. Carrying them is a protocol change, so the docblock now states both cases and what they currently project, and there is a test asserting `model.revivals` is `[]` whose message says that is also what a search matching nothing produces.

The real fix belongs with the relationship-state work, and I have written both this and the `not-loaded` reference loss onto CS-12622, which already specifies an explicit link-state union carrying a stable reference.

File: packages/host/app/lib/boxel-projection.ts (hunk 6 above)
Comment (habdelra) [reply packages/host/app/lib/boxel-projection.ts]: [Claude Code 🤖] Agreed, and the drift you spotted was real — filtering non-strings and mapping empty to null were both mine, not main's. Taken the export route in d69d0b43c4 rather than deferring it.

`isThemeCard` / `themeCss` / `hasTheme` / `themeId` / `getCssImports` are now module-scope exports of `field-component.gts` (they captured nothing from `getBoxComponent`, so lifting them was mechanical), re-exported from `card-api` the way `resolveFieldConfiguration` is. `captureTheme` calls them, and `cssImportsOf` and `fieldTypedAs` are gone — the capture no longer has a theme opinion of its own, only the record shape it puts main's answers into.

File: packages/host/app/lib/direct-boxel-runtime.ts (hunk 7 above)
Comment (habdelra) [reply packages/host/app/lib/direct-boxel-runtime.ts]: [Claude Code 🤖] Right — and I went with the named error rather than narrowing the registry, because the pipeline genuinely does describe field instances (`boxelKind` answers `'field'`, and `projectInstance` on one is meaningful). It is only the *document* form that is card-shaped. Fixed in d69d0b43c4: `serializeCard` checks `api.isCard` and throws naming both the handle and the class, instead of dying inside `isSingleCardDocument` with a shape dump that names neither.

File: packages/host/app/lib/direct-boxel-runtime.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (habdelra) [reply packages/host/app/lib/direct-boxel-runtime.ts]: [Claude Code 🤖] Both taken, in d69d0b43c4.

`projectInstance` now captures the instance half alone via a new `captureInstanceProjection` and calls `buildInstanceProjection` — no ancestry walk, no format resolution, no configuration functions evaluated, one normalize instead of three. `buildBoxelRenderRecord` stays as the all-three entry point.

And you are right that the header claim expired the moment `getFields` got its own read. It now says what one capture actually buys — a record that is internally coherent, so an instance's field list and its model come from a single read and never describe a state it was never in — and says explicitly that it is not a claim two operations agree, since each reads when it is called.

File: packages/host/app/lib/boxel-projection.ts (hunk 4 above)
Comment (backspace) [reply packages/host/app/lib/boxel-projection.ts:957]: [Claude Code 🤖] Both halves land.

`BigIntegerSerializer.serialize` is `val == null ? undefined : String(val)`, so the spelling matches the serializer exactly rather than approximately. And omission does reach the diagnostic — `watch`'s `!(key in target)` is true for a member `defineMember` never created, so the read reports the path. It also composes with the parity harness, which already carries `'absent'` as its own divergence reason, so a tier that omits a member and one that carries it part visibly.

One thing the change did that the reply does not mention. Dropping `?? null` from `captureValue` means an unset primitive field now projects as a *present member holding `undefined`* rather than `null` — `Contains.emptyValue` hands back the class's `[emptyValue]`, which is unassigned on `BaseDef` and so `undefined` for everything but Boolean. That gives the model the three-way split the `notData` docblock always claimed and nothing implemented: member absent (could not cross), present and `undefined` (unset), present with a value. Better than what I asked for — and it is the model's spelling for "empty" changing under every future consumer, with nothing in the suite reading an unset contains field. Worth a line in `captureModel`'s docblock or a pinned assertion, because the next reader will assume `null`.

File: packages/host/app/lib/boxel-projection.ts (hunk 5 above)
Comment (backspace) [reply packages/host/app/lib/boxel-projection.ts:548]: [Claude Code 🤖] Accepted as documentation, and the pin is honest rather than cementing: the test's own message says the record has no spelling for the difference and that a tier reading it shows "no results" for a query that has not started — so it records the gap instead of blessing it. `deepEqual(model.revivals, [])` also fails if the shape later changes to omit query fields from the model, which is the direction a real fix would take, so it will not quietly outlive the decision it documents.

The docblock states both whole-field cases and why neither fits RP-7.1's slot union. Nothing further from me here.

File: packages/host/app/lib/boxel-projection.ts (hunk 6 above)
Comment (backspace) [reply packages/host/app/lib/boxel-projection.ts]: [Claude Code 🤖] The lift checks out: the five bodies moved byte for byte, and their only free identifiers — `getField`, `CardDef`, `Theme` — were already module-scope imports of `field-component.gts`, so nothing was captured. `boxel-projection.ts` is the only consumer and reaches all five through `api.*`, so there is one spelling in the tree.

Two leftovers, both small.

The reply says the capture no longer has a theme opinion of its own, but `cssImports` still normalizes: `getCssImports` hands `card.cssImports` back untouched, and the capture maps an empty array to `null` and filters non-strings. Mapping `undefined` to `null` is the record's shape and has to happen somewhere — the other two are the opinion the thread was about. They also disagree with each other: `length > 0` is tested before the filter, so `[]` projects as `null` while `[123]` projects as `[]`.

And the removal left a double blank line at the old definition site, which prettier flags. Nothing in CI will tell you — `packages/base`'s `lint` script is `lint:hbs` plus the isUsed check, with no eslint or prettier pass, so this file's formatting is only ever enforced by the commit hook.

File: packages/host/app/lib/direct-boxel-runtime.ts (hunk 7 above)
Comment (backspace) [reply packages/host/app/lib/direct-boxel-runtime.ts]: [Claude Code 🤖] The named error is the right call, and it sits ahead of everything that can throw.

The cast the thread was about is still there, though. `isCard` is exported from `card-api` as `card is CardDef`, so after the guard `card` is narrowed and `serializeCard(model: CardDef, …)` takes it directly — `as never` has no job left, and while it stands the next change to that signature still will not be caught here. Drop it, or say what it is standing for.

File: packages/host/app/lib/direct-boxel-runtime.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (backspace) [reply packages/host/app/lib/direct-boxel-runtime.ts]: [Claude Code 🤖] Both halves land. `captureInstanceProjection` reads the same four members the inline object did, and `captureBoxelInstance` delegates to it rather than duplicating them — so the fix for a second-builder problem did not create one. The header now claims only what survives.

It did leave two exports with no caller. `captureBoxelInstance` is referenced nowhere in the tree — not by production code, not by a test — now that `captureBoxelType`, `captureInstanceProjection` and `captureBoxelFields` are each called directly; `buildBoxelRenderRecord` is down to the unit test. Both are plausible as the entry point a tier answering all three at once will want, but nothing says so, and an exported function with no caller reads as an oversight to whoever arrives next. Either drop them until a caller exists, or have each docblock say it is the all-three entry point and that nothing calls it yet.

**Review by backspace — COMMENTED** [review]
[Claude Code 🤖] Re-review of the response commit, scoped to the five open threads and to what each change implicates around it.

**Convergence holds and the gate is met: no blocking issues.** Every thread moved on the mechanism it named rather than its symptom, no fix landed on only one twin, and nothing new turned up above nit level. Severity is strictly decreasing: the two I called worth settling before a consumer exists — a value that could not cross being indistinguishable from an empty one, and a second read path for one instance — are both closed at the mechanism, and the second was closed without introducing the very duplication it was about.

Dispositions; detail stays in each thread:

1. `dataValue` / `captureValue` — **resolves it.** `String(bigint)` is exactly what the serializer writes, and omission does reach the diagnostic. One undocumented side effect: unset primitives now project as `undefined` rather than `null`.
2. Query-backed whole-field states — **resolves it as documentation.** The pinning test names the gap in its own assertion message, so it records current behaviour rather than blessing it.
3. `captureTheme` — **resolves it.** The lift captured nothing and the tree has one spelling. Two nits: the `cssImports` normalization is still the capture's own opinion, and the removal left a prettier-flagged blank line that no CI step for that package will catch.
4. `serializeCard` — **resolves the mechanism, not the cast.** The guard narrows, so `as never` is now removable.
5. `projectInstance` — **resolves both halves**, and orphans `captureBoxelInstance` (no caller anywhere) and `buildBoxelRenderRecord` (test-only).

Outside the threads: the description still says 13 integration tests and a bijection count dropping "from 104 to 91", where the suite now has 14 and the script records 88. It is the public record of the change, so worth a pass before merge.

**Review by backspace — APPROVED** [review]

File: packages/host/app/lib/boxel-projection.ts (hunk 4 above)
Comment (habdelra) [reply packages/host/app/lib/boxel-projection.ts:957]: [Claude Code 🤖] You are right that it was a semantic change I made without noticing, and pinning it turned up that my own account of it was too broad. Fixed in bd585d66a4.

A card served by a realm carries its unauthored primitives as `null`, so a store-loaded instance projects `null`, not `undefined` — the `undefined` case is narrower than "unset": it is a field the instance's data never carried at all. My first attempt at the test asserted on a realm-loaded card and failed for exactly that reason. It now materializes a document carrying one attribute and asserts the live instance and the model agree, and the docblock states which of the two cases yields which.

File: packages/host/app/lib/boxel-projection.ts (hunk 6 above)
Comment (habdelra) [reply packages/host/app/lib/boxel-projection.ts]: [Claude Code 🤖] Both fixed in bd585d66a4.

The `cssImports` inconsistency was real and worse than it looked: emptiness was tested *before* the filter, so a list of nothing but unauthored entries projected as `[]` while a list of none projected as `null`. Filtering first makes the two agree. I kept the filter and the null-mapping but split them in the comment — the null is the record's shape, the filter is what keeps the member the `string[]` it is declared to be, since a `containsMany` entry that was never authored reads as the empty value rather than as a URL.

And the blank line is gone. Your point about it is the more useful half: `packages/base`'s `lint` is `lint:hbs` plus the isUsed check, so nothing in CI formats that file and only the commit hook would have caught it.

File: packages/host/app/lib/direct-boxel-runtime.ts (hunk 7 above)
Comment (habdelra) [reply packages/host/app/lib/direct-boxel-runtime.ts]: [Claude Code 🤖] Dropped in bd585d66a4 — `isCard` narrows to `CardDef`, so `serializeCard` takes `card` directly and the cast had nothing left to stand for. Agreed on the reason it mattered more than the noise: while it stood, the next change to that signature would have gone unchecked at this call site.

File: packages/host/app/lib/direct-boxel-runtime.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (habdelra) [reply packages/host/app/lib/direct-boxel-runtime.ts]: [Claude Code 🤖] Took the first option for one and the second for the other, in bd585d66a4.

`captureBoxelInstance` is gone. It had no caller anywhere once the three captures were each called directly, and I would rather a tier that wants all three compose them explicitly than inherit a helper nobody had used.

`buildBoxelRenderRecord` stays, and its docblock now says what it is: the entry point for a caller that wants all three at once, with nothing on a render path calling it yet because each operation asks for the one record it answers with. Worth saying rather than dropping — it is the assembler's own composition, and removing it would leave the three builders with no statement anywhere that they are one pipeline.

File: packages/host/app/lib/boxel-projection-diagnostics.ts [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+ * for reading through. A view does not survive `structuredClone` — a Proxy
+ * never does — which is the same rule stated from the other side.
+ */
+export function observeMissingProjectionPaths<T>(
+  record: T,
+  context: MissingProjectionPathContext,
+): T {
+  if (config.environment === 'production') {
+    return record;
+  }
```
Comment (richardhjtan) [inline packages/host/app/lib/boxel-projection-diagnostics.ts:68]: [Claude Code 🤖] The header above this says "**It never ships.** In a production build this is the identity function, so no proxy exists and no read is intercepted." The second half is true; the first isn't, and they read as one claim.

`config.environment === 'production'` is a runtime read of the environment module, so nothing here is eliminated at build time: this whole file — the `watch` walk, the Proxy construction, `VALUE_PROBES`, the reporter — is bundled into the production build and the branch is evaluated on every call. What the guard buys is the behavior (no proxy is *created*, no read is *intercepted*), not the absence.

Two ways to make the header true, and either is fine:

- Strip it. `@embroider/macros` is already the idiom here — `isTesting` is imported from it in eight places under `packages/host/app` — and `macroCondition(!isDevelopingApp())` is build-time-replaced, so the branch and its dead side go away. Worth confirming against the host's build before relying on it.
- Or reword to what the guard actually holds: it is inert in production rather than absent from it.

The reason to bother with a comment-level thing: this file's whole argument for existing is that a development-only instrument is safe to add, and "it never ships" is the load-bearing sentence in that argument. Someone will quote it.

Non-blocking.

File: packages/host/app/lib/boxel-render-record.ts [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+ * render path does yet — each operation asks for the one record it answers
+ * with, which is why `describeBoxel`, `getFields` and `projectInstance` each
+ * call the builder for theirs — so today this composes them for the suite that
+ * holds the three to one input.
+ *
+ * It is one builder per record either way: a caller assembling a description
+ * here and a projection somewhere else would have two, which is the shape this
+ * file exists to make unnecessary.
+ */
+export function buildBoxelRenderRecord(
```
Comment (richardhjtan) [inline packages/host/app/lib/boxel-render-record.ts:92]: [Claude Code 🤖] The last leftover from the re-review that's still true: this is exported, and the only callers anywhere are four sites in `packages/host/tests/unit/rp-14-render-record-test.ts`. Nothing in `app/` reaches it.

Given the PR's own framing — one pipeline, and a second builder is the failure mode the whole design exists to prevent — a public entry point that only tests use is the seam a second builder would grow from: it is the assembler's front door, reachable without going through a capture. Either fold it into whatever the production path calls and let the test reach it through that, or keep it and say in the header that it is the assembler's test seam, so the next reader doesn't take a public export as an invitation.

Non-blocking; the ask is a decision about the module's surface, not a defect.

**Review by richardhjtan — COMMENTED** [review]
[Claude Code 🤖] Two things this pass went after, neither of which the existing exchange settles: whether the outstanding punch list from the re-review is still outstanding, and whether the diagnostic's own claims about itself hold. I did not re-open the five converged threads.

**The punch list is stale — four of the five leftovers are already fixed in this head, and the description numbers are now right.** Verified rather than inferred:

- `captureBoxelInstance` has no references anywhere in the tree, including its own definition — the orphan is gone, not just uncalled.
- There is no `as never` left in `direct-boxel-runtime.ts`.
- `npx prettier --check` is clean across `field-component.gts`, `card-api.gts`, `boxel-projection.ts`, `boxel-render-record.ts` and `direct-boxel-runtime.ts`, so the flagged blank line is gone.
- The description's counts match the tree exactly: 21 integration tests (the file has 21 `test(` calls), 10 on the assembler, 6 on the diagnostic, and 104 → 88 on the bijection.

Worth saying because anyone reading those threads to decide on merge would otherwise think there is work left in them.

**Two asks, both non-blocking.** One remaining leftover — `buildBoxelRenderRecord` is exported with no production caller — plus one thing the diagnostic asserts about itself that isn't true. Threads on each.

**One description error survives the pass.** It names the diagnostic entry point `watchProjectionPaths`; the module's only exported function is `observeMissingProjectionPaths`. That's the harder kind of stale reference, because a reader who searches for the name in the description finds nothing at all.

**Comment by habdelra** [comment]
not implementing this
